pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_controller.go

```diff
@@ -46,124 +54,225 @@ func (r *HTTPRouteReconciler) Reconcile(ctx context.Context, req kube_ctrl.Reque
+	if ns := ref.Namespace; ns != nil {
+		namespace = string(*ns)
+	}
 
-		for _, backend := range rule.BackendRefs {
-			ref := backend.BackendObjectReference
+	gateway := &gatewayapi.Gateway{}
 
-			destination, err := r.gapiToKumaRef(ctx, mesh, namespace, ref)
-			if err != nil {
-				return nil, err
-			}
+	if *ref.Group != gatewayapi.Group(gatewayapi.GroupName) || *ref.Kind != gatewayapi.Kind(reflect.TypeOf(gateway).Elem().Name()) {
+		return false, nil
+	}
 
-			backends = append(backends, &mesh_proto.GatewayRoute_Backend{
-				// Weight has a default of 1
-				Weight:      uint32(*backend.Weight),
-				Destination: destination,
-			})
+	if err := r.Client.Get(ctx, kube_types.NamespacedName{Namespace: namespace, Name: string(ref.Name)}, gateway); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			return false, nil
+		} else {
+			return false, err
 		}
+	}
+
+	class, err := getGatewayClass(ctx, r.Client, gateway.Spec.GatewayClassName)
+	if err != nil {
+		return false, err
+	}
+
+	return class.Spec.ControllerName == controllerName, nil
+}
 
-		var matches []*mesh_proto.GatewayRoute_HttpRoute_Match
+// reconcileLabelledObjectSet manages a set of owned objects based on labeling
+// with the owner key and indexed by the service tag and listener tag.
+func reconcileLabelledObject(
+	ctx context.Context,
+	client kube_client.Client,
+	owner kube_types.NamespacedName,
+	routeSpec *mesh_proto.GatewayRoute,
+) error {
+	// First we list which existing objects are owned by this route.
+	// We expect either 0 or 1 and depending on whether routeSpec is nil
+	// we either create an object or update or delete the existing
+	ownerLabelValue := fmt.Sprintf("%s-%s", owner.Namespace, owner.Name)
+	labels := kube_client.MatchingLabels{
+		ownerLabel: ownerLabelValue,
+	}
+
+	list := &mesh_k8s.GatewayRouteList{}
+	if err := client.List(ctx, list, labels); err != nil {
+		return err
+	}
+
+	if len(list.Items) > 1 {
+		return fmt.Errorf("expected either zero or one owned items")
+	}
+
+	var existing *mesh_k8s.GatewayRoute
+	if len(list.Items) == 1 {
+		existing = &list.Items[0]
+	}
 
-		for _, match := range rule.Matches {
-			kumaMatch, err := gapiToKumaMatch(match)
-			if err != nil {
-				return nil, errors.Wrap(err, "couldn't convert match")
+	if routeSpec == nil {
+		if existing != nil {
+			if err := client.Delete(ctx, existing); err != nil && !kube_apierrs.IsNotFound(err) {
+				return err
 			}
+		}
+		return nil
+	}
 
-			matches = append(matches, kumaMatch)
+	if existing != nil {
+		existing.Spec = routeSpec
+
+		if err := client.Update(ctx, existing); err != nil {
+			return errors.Wrap(err, "could not update GatewayRoute")
 		}
+		return nil
+	}
 
-		var filters []*mesh_proto.GatewayRoute_HttpRoute_Filter
+	route := &mesh_k8s.GatewayRoute{
+		ObjectMeta: kube_meta.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", ownerLabelValue),
+			Labels: map[string]string{
+				ownerLabel: ownerLabelValue,
+			},
+		},
+		Spec: routeSpec,
+	}
```

Comment:
Should we set the HTTPRoute as an owner reference here?

Reply:
We can't, that's part of the reason we need `reconcileLabelledObject` for deletion, see https://github.com/kumahq/kuma/issues/3517#issuecomment-996739915.